editor/plugins/visual_shader_editor_plugin.cpp

```diff
@@ -1483,9 +1483,36 @@ VisualShaderEditor::VisualShaderEditor() {
 	add_options.push_back(AddOption("Tangent", "Input", "Vertex", "VisualShaderNodeInput", TTR("'tangent' input parameter for vertex shader mode."), "tangent", VisualShaderNode::PORT_TYPE_VECTOR, VisualShader::TYPE_VERTEX));
 	add_options.push_back(AddOption("UV", "Input", "Vertex", "VisualShaderNodeInput", TTR("'uv' input parameter for vertex shader mode."), "uv", VisualShaderNode::PORT_TYPE_VECTOR, VisualShader::TYPE_VERTEX));
-	add_options.push_back(AddOption("UV2", "Input", "Vertex", "VisualShaderNodeInput", TTR("'uv2' input parameter for vertex shader mode."), "uv2", VisualShaderNode::PORT_TYPE_VECTOR, VisualShader::TYPE_VERTEX));
+	add_options.push_back(AddOption("UV2", "Input", "Vertex", "VisualShaderNodeInput", TTR("(GLES3 only) 'uv2' input parameter for vertex shader mode."), "uv2", VisualShaderNode::PORT_TYPE_VECTOR, VisualShader::TYPE_VERTEX));
 	add_options.push_back(AddOption("Vertex", "Input", "Vertex", "VisualShaderNodeInput", TTR("'vertex' input parameter for vertex shader mode."), "vertex", VisualShaderNode::PORT_TYPE_VECTOR, VisualShader::TYPE_VERTEX));
 
+	add_options.push_back(AddOption("Transform", "Input", "Vertex", "VisualShaderNodeInput", TTR("(Particles only) 'transform' ..."), "transform", VisualShaderNode::PORT_TYPE_TRANSFORM));
+	add_options.push_back(AddOption("Active", "Input", "Vertex", "VisualShaderNodeInput", TTR("(Particles only) 'active' ..."), "active", VisualShaderNode::PORT_TYPE_SCALAR));
+	add_options.push_back(AddOption("Velocity", "Input", "Vertex", "VisualShaderNodeInput", TTR("(Particles only) 'velocity' ..."), "velocity", VisualShaderNode::PORT_TYPE_VECTOR));
+	add_options.push_back(AddOption("Restart", "Input", "Vertex", "VisualShaderNodeInput", TTR("(Particles only) 'restart' ..."), "restart", VisualShaderNode::PORT_TYPE_SCALAR));
+	add_options.push_back(AddOption("Custom", "Input", "Vertex", "VisualShaderNodeInput", TTR("(Particles only) 'custom' ..."), "custom", VisualShaderNode::PORT_TYPE_VECTOR));
+	add_options.push_back(AddOption("CustomAlpha", "Input", "Vertex", "VisualShaderNodeInput", TTR("(Particles only) 'custom_alpha' ..."), "custom_alpha", VisualShaderNode::PORT_TYPE_SCALAR));
+	add_options.push_back(AddOption("Delta", "Input", "Vertex", "VisualShaderNodeInput", TTR("(Particles only) 'delta' ..."), "delta", VisualShaderNode::PORT_TYPE_SCALAR));
+	add_options.push_back(AddOption("Lifetime", "Input", "Vertex", "VisualShaderNodeInput", TTR("(Particles only) 'lifetime' ..."), "lifetime", VisualShaderNode::PORT_TYPE_SCALAR));
+	add_options.push_back(AddOption("Index", "Input", "Vertex", "VisualShaderNodeInput", TTR("(Particles only) 'index' ..."), "index", VisualShaderNode::PORT_TYPE_SCALAR));
+	add_options.push_back(AddOption("EmissionTransform", "Input", "Vertex", "VisualShaderNodeInput", TTR("(Particles only) 'em_transform' ..."), "emission_transform", VisualShaderNode::PORT_TYPE_TRANSFORM));
```

Comment:
I guess you can add the same "input parameter for x shader mode." to the descriptions - there is no complex english here.

Also dont forget to setup the shader modes - VisualShader::TYPE_VERTEX or (VisualShader::TYPE_FRAGMENT |  VisualShader::TYPE_LIGHT)

Reply:
okay